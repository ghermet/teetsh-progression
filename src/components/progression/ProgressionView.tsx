import type { ProgressionData } from '../../types/progression.types';
import { ProgressionFooter } from './ProgressionFooter';
import { ProgressionHeader } from './ProgressionHeader';
import { ProgressionTable } from './ProgressionTable';

/**
 * Main progression visualization component that orchestrates the display of educational progression data.
 *
 * Transforms hierarchical progression data (Matière → Domaine → Items) into a timeline-based
 * table visualization showing educational content organized by subjects and time periods.
 *
 * @param props - The complete progression data including metadata and content
 * @param props.name - The name/title of the progression
 * @param props.shortDescription - Brief description of the progression
 * @param props.date - Date when the progression was created
 * @param props.niveau - Educational level (e.g., "CP", "CE1")
 * @param props.periodes - Array of time periods with dates and colors
 * @param props.matieres - Array of subjects containing domains and learning items
 * @returns The complete progression view with header, table, and footer
 */
export function ProgressionView({
  name,
  shortDescription,
  date,
  niveau,
  periodes,
  matieres,
}: ProgressionData) {
  return (
    <section className="bg-white rounded-lg sm:rounded-xl overflow-hidden shadow-sm sm:shadow-md border border-slate-200">
      <ProgressionHeader
        name={name}
        shortDescription={shortDescription}
        niveau={niveau}
        date={date}
      />
      <ProgressionTable periodes={periodes} matieres={matieres} />
      <ProgressionFooter periodes={periodes} matieres={matieres} />
    </section>
  );
}
