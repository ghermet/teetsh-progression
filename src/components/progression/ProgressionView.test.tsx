/** Generated with CLAUDE CODE */

import { render, screen } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import { ProgressionView } from './ProgressionView';
import type { ProgressionData } from '../../types/progression.types';

describe('ProgressionView', () => {
  const mockProgressionData: ProgressionData = {
    id: 1,
    name: 'Test Progression',
    shortDescription: 'Test Description',
    date: '2024-01-15',
    niveau: 'CP',
    userId: 'user-1',
    nbOfUseLanding: 0,
    nbOfUseInApp: 0,
    schoolyearId: 'school-year-1',
    schoolId: 'school-1',
    programmationId: 'prog-1',
    columnWidth: 200,
    fontSize: '14px',
    view: 'timeline',
    invertedRowCol: false,
    createdAt: '2024-01-01',
    updatedAt: '2024-01-01',
    publishedAt: '2024-01-01',
    onePageMatiere: null,
    slug: 'test-progression',
    documentId: 'doc-1',
    periodes: [
      {
        id: '1',
        name: 'Période 1',
        color: 'blue',
        startDate: '2024-01-01',
        endDate: '2024-01-07',
        position: 1,
        programmationId: 'prog-1',
      },
      {
        id: '2',
        name: 'Période 2',
        color: 'green',
        startDate: '2024-01-08',
        endDate: '2024-01-14',
        position: 2,
        programmationId: 'prog-1',
      },
    ],
    matieres: [
      {
        id: 'matiere-1',
        name: 'Math',
        color: 'blue',
        position: 1,
        programmationId: 'prog-1',
        domaines: [
          {
            id: 'domaine-1',
            name: 'Numbers',
            color: 'blue',
            matiereId: 'matiere-1',
            items: [
              {
                id: 'item-1',
                value: 'Addition',
                y: 0,
                Sequence: null,
                domaineId: 'domaine-1',
                periodeId: '1',
                FicheDePrep: null,
                attachments: [],
              },
              {
                id: 'item-2',
                value: 'Subtraction',
                y: 1,
                Sequence: null,
                domaineId: 'domaine-1',
                periodeId: '2',
                FicheDePrep: null,
                attachments: [],
              },
            ],
          },
        ],
      },
      {
        id: 'matiere-2',
        name: 'French',
        color: 'green',
        position: 2,
        programmationId: 'prog-1',
        domaines: [
          {
            id: 'domaine-2',
            name: 'Reading',
            color: 'green',
            matiereId: 'matiere-2',
            items: [
              {
                id: 'item-3',
                value: 'Letters',
                y: 0,
                Sequence: null,
                domaineId: 'domaine-2',
                periodeId: '1',
                FicheDePrep: null,
                attachments: [],
              },
            ],
          },
        ],
      },
    ],
  };

  it('renders all progression components', () => {
    render(<ProgressionView {...mockProgressionData} />);

    expect(
      screen.getByRole('heading', { name: /Test Progression/i })
    ).toBeInTheDocument();
    expect(screen.getByRole('table')).toBeInTheDocument();
    expect(screen.getByRole('contentinfo')).toBeInTheDocument();
  });

  it('passes correct props to ProgressionHeader', () => {
    render(<ProgressionView {...mockProgressionData} />);

    expect(
      screen.getByRole('heading', { name: /Test Progression/i })
    ).toBeInTheDocument();
    expect(screen.getByText('Test Description')).toBeInTheDocument();
    expect(screen.getByText('CP')).toBeInTheDocument();
    expect(screen.getByText('2024')).toBeInTheDocument();
  });

  it('transforms matieres into topics correctly', () => {
    render(<ProgressionView {...mockProgressionData} />);

    // Should have 2 topics (2 domaines)
    const rows = screen.getAllByRole('row');
    // 1 header row + 2 body rows
    expect(rows).toHaveLength(3);
  });

  it('sorts matieres by position', () => {
    const dataWithUnsortedMatieres = {
      ...mockProgressionData,
      matieres: [
        {
          ...mockProgressionData.matieres[1], // French with position 2
          position: 2,
        },
        { ...mockProgressionData.matieres[0], position: 1 }, // Math with position 1
      ],
    };

    render(<ProgressionView {...dataWithUnsortedMatieres} />);

    const rows = screen.getAllByRole('rowheader');
    // First row should be Math, second should be French
    expect(rows[0]).toHaveTextContent('Math');
    expect(rows[1]).toHaveTextContent('French');
  });

  it('handles empty matieres', () => {
    const dataWithEmptyMatieres = {
      ...mockProgressionData,
      matieres: [],
    };

    render(<ProgressionView {...dataWithEmptyMatieres} />);

    const rows = screen.getAllByRole('row');
    // 1 header row
    expect(rows).toHaveLength(1);
  });

  it('handles matieres with no items', () => {
    const dataWithEmptyItems = {
      ...mockProgressionData,
      matieres: [
        {
          ...mockProgressionData.matieres[0],
          domaines: [
            {
              ...mockProgressionData.matieres[0].domaines[0],
              items: [],
            },
          ],
        },
      ],
    };

    render(<ProgressionView {...dataWithEmptyItems} />);

    const rows = screen.getAllByRole('row');
    // 1 header row + 1 body row
    expect(rows).toHaveLength(2);
    expect(screen.queryByRole('listitem')).not.toBeInTheDocument();
  });

  it('has correct CSS classes and structure', () => {
    const { container } = render(<ProgressionView {...mockProgressionData} />);

    const section = container.querySelector('section');
    expect(section).toHaveClass(
      'bg-white',
      'rounded-lg',
      'overflow-hidden',
      'shadow-sm',
      'border',
      'border-slate-200'
    );
  });

  it('matches snapshot', () => {
    const { container } = render(<ProgressionView {...mockProgressionData} />);
    expect(container).toMatchSnapshot();
  });
});
