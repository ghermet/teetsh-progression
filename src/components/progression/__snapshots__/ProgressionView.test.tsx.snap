// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressionView > matches snapshot 1`] = `
<div>
  <section
    class="bg-white rounded-lg sm:rounded-xl overflow-hidden shadow-sm sm:shadow-md border border-slate-200"
  >
    <header
      class="bg-slate-100 px-4 sm:px-6 py-6 sm:py-8 border-b border-slate-200"
    >
      <div
        class="flex flex-col sm:flex-row sm:items-center sm:justify-between gap-4 sm:gap-0"
      >
        <div
          class="flex-1 min-w-0"
        >
          <div
            class="flex items-center justify-between flex-wrap gap-1"
          >
            <h1
              class="text-xl md:text-2xl lg:text-3xl text-slate-800 font-medium"
            >
              Test Progression
            </h1>
            <div
              class="flex items-baseline gap-2 text-slate-600 text-sm sm:text-base flex-shrink-0"
            >
              <strong
                class="text-slate-800"
              >
                CP
              </strong>
              <span
                class="hidden sm:inline"
              >
                •
              </span>
              <span
                class="text-slate-700"
              >
                2024
              </span>
            </div>
          </div>
          <p
            class="text-sm md:text-base lg:text-lg text-slate-600 mt-1"
          >
            Test Description
          </p>
        </div>
      </div>
    </header>
    <div
      class="relative overflow-x-auto"
    >
      <div
        class="min-w-max"
      >
        <table
          class="w-full border-collapse"
        >
          <caption
            class="sr-only"
          >
            Tableau de progression des matières par périodes
          </caption>
          <thead>
            <tr>
              <th
                class="sticky left-0 z-10 w-48 sm:w-56 md:w-64 bg-slate-50 border-r border-slate-200 h-12 px-2 sm:px-3 text-center"
                scope="col"
              >
                <span
                  class="text-slate-700 text-xs sm:text-sm font-medium"
                  data-testid="matieres-header"
                >
                  <span
                    class="hidden sm:inline"
                  >
                    Matières
                  </span>
                  <span
                    class="sm:hidden"
                  >
                    Mat.
                  </span>
                </span>
              </th>
              <th
                class="w-48 sm:w-56 md:w-64 border-r border-slate-200 text-center px-1 bg-slate-50 select-none"
                scope="col"
                style="background: var(--color-blue);"
                title="Période 1 du 01 janvier au 07 janvier"
              >
                <div
                  class="text-slate-700 text-xs sm:text-sm font-medium"
                  data-testid="periode-name"
                >
                  Période 1
                </div>
                <div
                  class="text-slate-800 text-xs hidden sm:block"
                >
                  01 janvier
                   - 
                  07 janvier
                </div>
                <div
                  class="text-slate-800 text-xs sm:hidden"
                >
                  01 janv.
                </div>
              </th>
              <th
                class="w-48 sm:w-56 md:w-64 border-r border-slate-200 text-center px-1 bg-slate-50 select-none"
                scope="col"
                style="background: var(--color-green);"
                title="Période 2 du 08 janvier au 14 janvier"
              >
                <div
                  class="text-slate-700 text-xs sm:text-sm font-medium"
                  data-testid="periode-name"
                >
                  Période 2
                </div>
                <div
                  class="text-slate-800 text-xs hidden sm:block"
                >
                  08 janvier
                   - 
                  14 janvier
                </div>
                <div
                  class="text-slate-800 text-xs sm:hidden"
                >
                  08 janv.
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th
                class="sticky left-0 z-10 flex flex-col w-48 sm:w-56 md:w-64 h-44 bg-slate-50 border-r border-t border-slate-200 px-2 sm:px-3 py-2 text-center select-none"
                scope="row"
              >
                <div
                  class="flex flex-col items-center gap-1 sm:gap-2 w-full min-h-0"
                >
                  <span
                    class="w-2 h-2 sm:w-3 sm:h-3 rounded-sm flex-shrink-0"
                    style="background: var(--color-blue);"
                  />
                  <span
                    class="text-slate-800 text-xs sm:text-sm font-medium text-center leading-tight overflow-hidden line-clamp-2"
                    data-testid="domaine-name"
                    title="Numbers"
                  >
                    Numbers
                  </span>
                  <span
                    class="text-slate-500 text-xs font-light text-center leading-tight overflow-hidden line-clamp-2"
                    data-testid="matiere-name"
                    title="Math"
                  >
                    Math
                  </span>
                </div>
                <div
                  class="text-slate-500 text-xs mt-auto"
                  data-testid="domaine-count"
                >
                  <span
                    class="hidden sm:inline"
                  >
                    2
                     semaines
                  </span>
                  <span
                    class="sm:hidden"
                  >
                    2
                  </span>
                </div>
              </th>
              <td
                class="min-w-48 sm:min-w-56 md:min-w-64 border-r border-slate-200 align-top p-1 overflow-auto bg-slate-50"
              >
                <ul
                  class="flex flex-col gap-1"
                >
                  <li
                    class="px-1.5 sm:px-2 py-1 rounded-md shadow-md transition-all min-h-6 text-slate-800 text-xs font-medium leading-tight break-words w-full select-none touch-manipulation"
                    style="background: var(--color-blue);"
                  >
                    Addition
                  </li>
                </ul>
              </td>
              <td
                class="min-w-48 sm:min-w-56 md:min-w-64 border-r border-slate-200 align-top p-1 overflow-auto bg-white"
              >
                <ul
                  class="flex flex-col gap-1"
                >
                  <li
                    class="px-1.5 sm:px-2 py-1 rounded-md shadow-md transition-all min-h-6 text-slate-800 text-xs font-medium leading-tight break-words w-full select-none touch-manipulation"
                    style="background: var(--color-blue);"
                  >
                    Subtraction
                  </li>
                </ul>
              </td>
            </tr>
            <tr>
              <th
                class="sticky left-0 z-10 flex flex-col w-48 sm:w-56 md:w-64 h-44 bg-slate-50 border-r border-t border-slate-200 px-2 sm:px-3 py-2 text-center select-none"
                scope="row"
              >
                <div
                  class="flex flex-col items-center gap-1 sm:gap-2 w-full min-h-0"
                >
                  <span
                    class="w-2 h-2 sm:w-3 sm:h-3 rounded-sm flex-shrink-0"
                    style="background: var(--color-green);"
                  />
                  <span
                    class="text-slate-800 text-xs sm:text-sm font-medium text-center leading-tight overflow-hidden line-clamp-2"
                    data-testid="domaine-name"
                    title="Reading"
                  >
                    Reading
                  </span>
                  <span
                    class="text-slate-500 text-xs font-light text-center leading-tight overflow-hidden line-clamp-2"
                    data-testid="matiere-name"
                    title="French"
                  >
                    French
                  </span>
                </div>
                <div
                  class="text-slate-500 text-xs mt-auto"
                  data-testid="domaine-count"
                >
                  <span
                    class="hidden sm:inline"
                  >
                    1
                     semaines
                  </span>
                  <span
                    class="sm:hidden"
                  >
                    1
                  </span>
                </div>
              </th>
              <td
                class="min-w-48 sm:min-w-56 md:min-w-64 border-r border-slate-200 align-top p-1 overflow-auto bg-slate-50"
              >
                <ul
                  class="flex flex-col gap-1"
                >
                  <li
                    class="px-1.5 sm:px-2 py-1 rounded-md shadow-md transition-all min-h-6 text-slate-800 text-xs font-medium leading-tight break-words w-full select-none touch-manipulation"
                    style="background: var(--color-green);"
                  >
                    Letters
                  </li>
                </ul>
              </td>
              <td
                class="min-w-48 sm:min-w-56 md:min-w-64 border-r border-slate-200 align-top p-1 overflow-auto bg-white"
              >
                <ul
                  class="flex flex-col gap-1"
                />
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <footer
      class="bg-slate-50 border-t border-slate-200 px-3 sm:px-4 py-2 sm:py-3"
    >
      <div
        class="flex items-center justify-between gap-4"
      >
        <div
          class="flex items-center gap-2 text-slate-600 text-xs sm:text-sm"
        >
          <span
            class="flex items-center gap-1"
          >
            <span
              class="font-medium"
            >
              2
            </span>
            <span>
              semaine
              s
            </span>
          </span>
          <span
            class="text-slate-400"
          >
            •
          </span>
          <span
            class="flex items-center gap-1"
          >
            <span
              class="font-medium"
            >
              2
            </span>
            <span>
              matière
              s
            </span>
          </span>
        </div>
        <div
          class="touch-only text-xs text-slate-500 bg-white/80 px-2 py-1 rounded-md shadow-sm pointer-events-none"
          role="alert"
        >
          ← Glissez pour défiler →
        </div>
      </div>
    </footer>
  </section>
</div>
`;
